s3/s3etag/s3etag.go

```diff
@@ -0,0 +1,141 @@
+// Package s3etag implements the AWS ETag hash as described in
+//
+// http://permalink.gmane.org/gmane.comp.file-systems.s3.s3tools/583
+package s3etag
+
+import (
+	"crypto/md5"
+	"encoding/binary"
+	"hash"
+
+	"github.com/ncw/rclone/s3/s3sizes"
+)
+
+const (
+	hashSize          = 32
+	partCountSize     = 4
+	size              = hashSize + partCountSize
+	hashReturnedError = "hash function returned error"
+)
+
+type digest struct {
+	n           int64 // bytes written into blockHash so far
+	blockHash   hash.Hash
+	totalHash   hash.Hash
+	sumCalled   bool
+	writtenMore bool
+	partSize    int64
+	partCount   uint16
+}
+
+// New returns a new hash.Hash computing the S3ETag checksum.
+func New(size int64) hash.Hash {
+	partSize := s3sizes.PartSize(size)
```

Comment:
This is going to check files which were uploaded by rclone just fine, but what about other programs which use a different partition size algorithm.  The really needs the number of partitions passed in here so it can be read out of the etag...

Reply:
Agreed, this is not a panacea, it does not support large files uploaded by a 3rd-party.   Note that this is not worse than S3's MD5 today, which does not support *any* large files.

In terms of fully defining the ETag algo, I don't think a single filesize parameter *OR* the single partition-count parameter is enough.  (For partition count, imagine ETagging a file twice, with slightly different partition sizes, so that the partition count is the same but the ETags are otherwise different.)  To fully parameterize the ETag algorithm, we need an ordered list of partition sizes (or equivalent).

Once the algorithm is fully specified, `(crypt) check` would then need to fetch and use the algorithm parameters while calculating the local ETag.  Passing info like this, during a check, is not currently supported by `(crypt) check`.  (See my comment above about enhancing the hash handshaking.)

Instead, I address only the use-case you mention: upload and check with rclone.  By addressing the restricted use-case, I can treat the parameters as hardcoded into rclone's `PartSize` method (so that they don't need to be hung on a file as metadata, nor passed around during a `(crypt) check`).

Reply:
> Agreed, this is not a panacea, it does not support large files uploaded by a 3rd-party. Note that this is not worse than S3's MD5 today, which does not support any large files.

However it will error when trying to download files uploaded by a 3rd party program :-(

> In terms of fully defining the ETag algo, I don't think a single filesize parameter OR the single partition-count parameter is enough

Yes I think it need either (assume all partitions the same size, length of file, number of partitions) or (list of all partitions and size)